### PR TITLE
Add release workflow for litani

### PR DESCRIPTION
**Pull request by ronakfof**
Tests
[Example Release](https://github.com/ronakfof/litani/releases/tag/litani-1.23.4) - [[workflow](https://github.com/ronakfof/litani/runs/5147017859?check_suite_focus=true)]
[Deb package](https://github.com/ronakfof/litani/releases/download/litani-1.23.4/litani-1.23.4.deb) - [[workflow](https://github.com/ronakfof/litani/runs/5147020743?check_suite_focus=true)]
[Example PR to tap](https://github.com/ronakfof/homebrew-litani/pull/11) - [[workflow](https://github.com/ronakfof/litani/runs/5147020632?check_suite_focus=true)]

*Description of changes:*

This workflow gets triggered when a tag is pushed of format `litani-*`.
Bot creates a release for that tag with a boilerplate description.
If a release is created successfully, it also creates a `debian` package for Litani on `ubuntu-20.04`, tests it and uploads it as an additional asset to the release and also create a PR to `aws/tap` with updated `version`, `tag` and `commit`.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: .github/workflows/release.yaml [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        LITANI_VERSION: ${{ needs.perform-release.outputs.version }}
+      steps:
+        - uses: actions/checkout@v2
+          with:
+            submodules: recursive
+        - name: Fetch dependencies
+          run: sudo apt-get install -y mandoc scdoc ninja-build
+        - name: make dir structure
+          run: mkdir -p litani-${{ env.LITANI_VERSION }}/{DEBIAN,usr/{bin,libexec/litani,share/{doc/{litani},man/{man1,man5,man7}}}}
```
Comment (angelonakos) [inline .github/workflows/release.yaml]: Do you think that the `mkdir -p` statement for:

`usr/{bin,libexec/litani,share/{doc/{litani},man/{man1,man5,man7}}}`

needs to be broken up in the sake of readability?

File: .github/workflows/release.yaml [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+          run: |
+            touch litani-${{ env.LITANI_VERSION }}/DEBIAN/control
+            cat << EOF > litani-${{ env.LITANI_VERSION }}/DEBIAN/control
+            Package: Litani
+            Version: ${{ env.LITANI_VERSION }}
+            Architecture: amd64
+            Depends: ninja-build, gnuplot, graphviz, python3.9, python3-jinja2, python3-voluptuous
+            Maintainer: Kareem Khazem <karkhaz@amazon.co.uk>
+            Description: A program that provides platform-independent job control.
+              AWS Build Accumulator collects build jobs from multiple sources before executing them concurrently. It provides platform-independent job control (timeouts, return code control) and an output format that is easy to render into reports (for example, using the built-in renderer). AWS Build Accumulator shines where your project uses multiple different build systems or requires a unified interface describing heterogeneous build jobs.
```
Comment (angelonakos) [inline .github/workflows/release.yaml]: Can this line be wrapped at a column width of 80?

File: .github/workflows/release.yaml [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+            mv litani-${{ env.LITANI_VERSION }}/usr/libexec/litani/doc/out/man/*.5 litani-${{ env.LITANI_VERSION }}/usr/share/man/man5
+            mv litani-${{ env.LITANI_VERSION }}/usr/libexec/litani/doc/out/man/*.7 litani-${{ env.LITANI_VERSION }}/usr/share/man/man7
+        - name: Remove unnecessary files
+          run: |
+            rm -rf litani-${{ env.LITANI_VERSION }}/usr/libexec/litani/doc
+            sudo rm -rf /usr/libexec/litani
+        - name: Create .deb package
+          id: create_packages
+          run: |
+            sudo dpkg-deb --build --root-owner-group litani-${{ env.LITANI_VERSION }}
```
Comment (angelonakos) [inline .github/workflows/release.yaml]: are `sudo` and `--root-owner-group` necessary here?

File: .github/workflows/release.yaml [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+            echo "::set-output name=deb_package_name::$deb_package_name"
+        - name: Setup deb package
+          run: |
+            sudo apt-get update
+            sudo apt install -y ./litani-${{ env.LITANI_VERSION }}.deb
+        - name: Test deb package
+          run: |
+            litani -h
+            man litani
+            litani init --project-name nyx --stages codegen verify check
```
Comment (angelonakos) [inline .github/workflows/release.yaml]: These 3 stage names can be removed.

File: .github/workflows/release.yaml [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+            litani add-job --command 'which brew' --pipeline-name 'siesta' --ci-stage codegen --phony-outputs phony_1
+            litani add-job --command 'which python3' --pipeline-name 'siesta' --ci-stage verify --inputs phony_1 --phony-outputs phony_2
+            litani add-job --command 'which docker' --pipeline-name 'siesta' --ci-stage check --inputs phony_2  --phony-outputs phony_3
```
Comment (angelonakos) [inline .github/workflows/release.yaml]: Is there a possibility `docker` might ever be missing?

File: .github/workflows/release.yaml [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+            litani run-build
+        - name: Get release
+          id: get_release_info
+          uses: bruceadams/get-release@v1.2.3
```
Comment (angelonakos) [inline .github/workflows/release.yaml]: Is the upload_url  the only piece of information needed from this release? Can it be fetched differently without having to invoke the following GitHub action?

https://github.com/bruceadams/get-release

File: .github/workflows/release.yaml [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+name: Create Release and package
+
+jobs:
+  get-version-information:
+    name: Get Version Information
+    runs-on: ubuntu-20.04
+    outputs:
+      version: ${{ steps.split-version.outputs._1 }}
+    steps:
+      - uses: jungwinter/split@v1
```
Comment (karkhaz) [inline .github/workflows/release.yaml]: Why not v2?

I guess for each of these externally-authored steps that we use, we should audit them and then stick to a particular version, right? What about security issues?

File: .github/workflows/release.yaml [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+        uses: actions/checkout@v2
+      - name: Create release
+        uses: actions/create-release@v1
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          LITANI_VERSION: ${{ needs.get-version-information.outputs.version }}
+        with:
+          tag_name: litani-${{ env.LITANI_VERSION }}
+          release_name: litani-${{ env.LITANI_VERSION }}
+          body: |
```
Comment (karkhaz) [inline .github/workflows/release.yaml]: Is there any way we can include the changelog entry in here?

File: .github/workflows/release.yaml [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+        - name: Setup directory for deb package
+          run: |
+            sudo apt-get install -y mandoc scdoc ninja-build
+            mkdir -p litani-${{ env.LITANI_VERSION }}/{DEBIAN,usr/{bin,libexec/litani,share/{doc/litani,man/{man1,man5,man7}}}}
+            touch litani-${{ env.LITANI_VERSION }}/DEBIAN/control
+            cat << EOF > litani-${{ env.LITANI_VERSION }}/DEBIAN/control
+            Package: Litani
+            Version: ${{ env.LITANI_VERSION }}
+            Architecture: amd64
+            Depends: ninja-build, gnuplot, graphviz, python3-jinja2, python3-voluptuous
```
Comment (karkhaz) [inline .github/workflows/release.yaml]: Let's not install voluptuous. I'd like to keep this as optional.

File: .github/workflows/release.yaml (hunk 5 above)
Comment (karkhaz) [reply .github/workflows/release.yaml]: Same with brew.

I would prefer much more trivial commands like `/usr/bin/true`. And probably there only needs to be one job.

**Review by karkhaz — CHANGES_REQUESTED** [review]
Looking forward to this. My main question is whether we can integrate the changelog into the release notes.

Maybe the right thing to do here is: instead of this workflow triggering on `litani-*` tag, it would trigger on any commit that gets pushed with an updated changelog. In that case, it would take the release notes from the changelog, and the workflow itself would create the release tag.

So from a user perspective, when we want to create a new release, we add to the CHANGELOG and push that change to the `develop` branch, and this workflow would merge the `release` branch into that commit and tag it as the next release.

That might be more complicated though because the workflow itself would have to figure out the release number... But I'm quite keen for the release notes to end up in the tag message.

File: .github/workflows/release.yaml (hunk 1 above)
Comment (ronakfof) [reply .github/workflows/release.yaml]: The reason i am using `mkdir -p` is to not add unnecessary code to the workflow which will increase the length of code as well as be confusing to understand the directory structure which is way better with this one statement.

File: .github/workflows/release.yaml (hunk 2 above)
Comment (ronakfof) [reply .github/workflows/release.yaml]: Resolved

File: .github/workflows/release.yaml (hunk 3 above)
Comment (ronakfof) [reply .github/workflows/release.yaml]: `sudo` is required here.
And the `--root-owner-group` flag makes all deb package content owned by the root user, which is the standard way to go. Without such flag, all files and folders would be owned by the user, which might not exist in the system the deb package would be installed to.

File: .github/workflows/release.yaml (hunk 4 above)
Comment (ronakfof) [reply .github/workflows/release.yaml]: Resolved

File: .github/workflows/release.yaml (hunk 5 above)
Comment (ronakfof) [reply .github/workflows/release.yaml]: Resolved

File: .github/workflows/release.yaml (hunk 6 above)
Comment (ronakfof) [reply .github/workflows/release.yaml]: A standard way is to use a GitHub Action to do so as i have seen while investigating. We can create our own to do so but this one is widely used.

File: .github/workflows/release.yaml (hunk 7 above)
Comment (ronakfof) [reply .github/workflows/release.yaml]: We use this one for a very simple change and I guess version won't matter here but we can definitely keep it updated.
Also I saw this being used in one of the `cbmc` workflows for the same.

File: .github/workflows/release.yaml (hunk 8 above)
Comment (ronakfof) [reply .github/workflows/release.yaml]: As discussed in the meetings, we will be triggering this workflow only on tag push to release branch.

File: .github/workflows/release.yaml (hunk 9 above)
Comment (ronakfof) [reply .github/workflows/release.yaml]: Resolved

File: .github/workflows/release.yaml (hunk 7 above)
Comment (angelonakos) [reply .github/workflows/release.yaml]: What about:
> I guess for each of these externally-authored steps that we use, we should audit them and then stick to a particular version, right? What about security issues?

File: .github/workflows/release.yaml (hunk 6 above)
Comment (angelonakos) [reply .github/workflows/release.yaml]: I'm fine with sticking with this widely used step. Thanks

File: .github/workflows/release-packages.yaml [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+  release:
+    types: [created]
+
+name: Upload additional release assets
+jobs:
+  ubuntu-20_04-package:
+    name: Generate ubuntu-20.04 debian package
+    runs-on: ubuntu-20.04
+    env:
+      GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
Comment (angelonakos) [inline .github/workflows/release-packages.yaml:11]: Should this be `RELEASE_CI_ACCESS_TOKEN`?

Comment (ronakfof) [reply .github/workflows/release-packages.yaml:11]: Bot is not involved in creation of the `debian` package sop we don't need it here. We can use the default token here.

File: .github/workflows/release.yaml (hunk 7 above)
Comment (ronakfof) [reply .github/workflows/release.yaml]: Resolved

File: .github/workflows/release-packages.yaml (hunk 10 above)
Comment (angelonakos) [reply .github/workflows/release-packages.yaml:11]: Got it.

Comment (angelonakos) [reply .github/workflows/release-packages.yaml:11]: https://docs.github.com/en/actions/security-guides/automatic-token-authentication#about-the-github_token-secret for those not familiar

**Review by angelonakos — APPROVED** [review]

**Review by karkhaz — APPROVED** [review]
This is awesome :)

**Pull request merged**